lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs

```diff
@@ -897,59 +903,116 @@ _evaluateMinimumFee
     :: Cardano.ProtocolParameters
     -> SealedTx
     -> Maybe Coin
-_evaluateMinimumFee pp tx =
-    fromCardanoLovelace <$> minFee
+_evaluateMinimumFee pp tx = do
+    anyShelleyTx <- inAnyShelleyBasedEra (cardanoTx tx)
+    pure $ withShelleyBasedBody anyShelleyTx $ \_era body ->
+        fromCardanoLovelace $
+            Cardano.evaluateTransactionFee
+                pp
+                body
+                (estimateNumberOfWitnesses body)
+                0
+
+-- | Estimate the size of the transaction (body) when fully signed.
+_estimateSignedTxSize
+    :: Cardano.IsShelleyBasedEra era
+    => Cardano.ProtocolParameters
+    -> Cardano.TxBody era
+    -> TxSize
+_estimateSignedTxSize pparams body =
+    let
+        nWits :: Word
+        nWits = estimateNumberOfWitnesses body
+
+        -- Hack which allows us to rely on the ledger to calculate the size of
+        -- witnesses:
+        feeOfWits :: Natural
+        feeOfWits = minfee nWits - minfee 0
```

Comment:
Suggestion: use `Coin` here, along with the `Coin.difference` function.

```suggestion
        feeOfWits :: Coin
        feeOfWits = minfee nWits `Coin.difference` minfee 0
```

This would also require you to change `minFee` so that it doesn't unwrap the `Coin` value, and also change `sizeOfWits` so that it does.

Advantage: the type signature declares the intent more clearly: that this is a lovelace value, rather than some other sort of natural number.

Reply:
Needed to also add a:
```
coinQuotRem :: Coin -> Coin -> (Natural, Natural)
coinQuotRem (Coin p) (Coin q) = quotRem p q
```
but it turned out pretty neat 👍